### PR TITLE
[karma] Add/update multiple types

### LINKED ISSUE #52282: Karma Server constructor does not accept Config as parameter
Authors: @peterblazejewicz

The Karma Server constructor should also be allowed a Config object (from config.js) as first parameter.

### PR DESCRIPTION
**Pull request by devoto13**
This PR makes several small updates, each in a separate commit. Please review commit by commit for convenience. Links to the supporting code/documentation are also present in each separate commit.

For context: this PR should fix various problems with @types/karma found in https://github.com/angular/angular-cli/pull/20485.

Fixes #52282

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/karma/index.d.ts [hunk 1]
```diff
@@ -700,6 +692,31 @@ export interface BrowserConsoleLogOptions {
...
+     * @deprecated Will become a default in the next major release.
+     */
+    throwErrors: boolean;
+}
+
 export namespace config {
     function parseConfig(configFilePath: string, cliOptions: ConfigOptions): Config;
+    // The return value is technically not correct as user can pass options and set promiseConfig to false.
+    // But from the pragmatic point of view, nobody should pass it like that, so it is better to provide a useful,
+    // although not technically correct return type.
```
Comment (43081j) [inline types/karma/index.d.ts:716]: The default is actually false, so we should really support the correct return type if we can. You can have two overloads, one which has it set to true and the other with it set to false. Or a conditional return type. Whichever is easier

Comment (peterblazejewicz) [reply types/karma/index.d.ts:716]: someting like (pseudo-code)
...
function parseConfig(configFilePath: string, cliOptions: ConfigOptions): Config;
function parseConfig(configFilePath: string, cliOptions: ConfigOptions & { promiseConfig: true }): Promise<Config>;
```
We can't bind to `promiseConfig?: true;
```
as there is probably existing code that uses `false`, like:
https://github.com/karma-runner/karma/pull/3660/files#diff-37b88258822cb207e2caa1a88d4fbc057924da72c5580d1f39b78349214ab622R58

File: types/karma/index.d.ts [hunk 2]
```diff
@@ -700,6 +692,31 @@ export interface BrowserConsoleLogOptions {
...
 }
 
+interface ParseOptions {
+    /**
+     * When true, the return value of the function is a Promise of Config instead of Config.
+     * Should be set to true to support async Karma configuration file.
+     *
+     * @deprecated Will become a default in the next major release.
+     */
+    promiseConfig: boolean;
```
Comment (peterblazejewicz) [inline types/karma/index.d.ts:702]:
```suggestion
    promiseConfig?: boolean;
```

File: types/karma/index.d.ts [hunk 3]
```diff
@@ -700,6 +692,31 @@ export interface BrowserConsoleLogOptions {
...
+
+    /**
+     * When true, function will throw error or return rejected Promise instead of calling process.exit(1).
+     *
+     * @deprecated Will become a default in the next major release.
+     */
+    throwErrors: boolean;
```
Comment (peterblazejewicz) [inline types/karma/index.d.ts:709]:
```suggestion
    throwErrors?: boolean;
```
should this be optional? I cannot see sample snip using this with `false` (contrary to `promiseConfig`), always as `true` or comparing against truthy/true (=== or if(...))

File: types/karma/index.d.ts (hunk 2 above)
Comment (devoto13) [reply types/karma/index.d.ts]: Yes, fixed!

File: types/karma/index.d.ts (hunk 1 above)
Comment (devoto13) [reply types/karma/index.d.ts]: Well, that code is only in Karma itself for backward compatibility. I don't think any Karma consumers will actually do that as the behavior is deprecated and there is no reason to use it over omitting the parameters altogether. But in any case, the main reason was that I didn't know how to make the correct return type 😄 Thanks for the tip @peterblazejewicz! I've updated the PR.

**Review by peterblazejewicz — APPROVED** [review]
LGMT!
@devoto13 thx!

**Review by 43081j — APPROVED** [review]

**Comment by devoto13** [comment]
Ready to merge

**Comment by erikhorlings** [comment]
Hey all,
Just wanted to say thanks for the work you put in.

**Pull request merged**